src/app/services/connectivity/connectivity.service.ts

```diff
@@ -1,30 +1,51 @@
 import { Injectable } from '@angular/core';
-import { BehaviorSubject, Observable } from 'rxjs';
+
+import { ConnectableObservable, Observable } from 'rxjs';
+import { multicast, publishReplay } from 'rxjs/operators';
+
 
 @Injectable({
   providedIn: 'root'
 })
 export class ConnectivityService {
-  private monitor: Observable<boolean>;
-  private subject: BehaviorSubject<boolean>;
-
-  constructor() {
-    this.monitor = new Observable((connectivityObserver) => {
-      // map 'offline' and 'online' window events to boolean 'connnectivity state'
-      window.addEventListener('offline', () => {
-        connectivityObserver.next(false);
-      });
-      window.addEventListener('online', () => {
-        connectivityObserver.next(true);
-      });
-    });
-    // obtain initial connectivity state from navigator.onLine property
-    this.subject = new BehaviorSubject(navigator.onLine);
-    this.monitor.subscribe(this.subject);
+  public static readonly OFFLINE_EVENT = 'offline';
+  public static readonly ONLINE_EVENT = 'online';
+  private connectivityMonitor: Observable<boolean>;
+  private connectivityObserver: ConnectableObservable<boolean>;
+
+  constructor() { }
+
+  initConnectivityObserver() {
+    this.connectivityMonitor = new Observable(
+      (connectivityStateObserver) => {
+        const offlineEventListener = () => {
+          connectivityStateObserver.next(false); // Browser offline -> connectivity: false
+        };
+        const onlineEventListener = () => {
+          connectivityStateObserver.next(true); // Browser online -> connectivity: true
+        };
+
+        window.addEventListener(ConnectivityService.OFFLINE_EVENT, offlineEventListener);
+        window.addEventListener(ConnectivityService.ONLINE_EVENT, onlineEventListener);
+
+        // Emit the current connectivity
+        connectivityStateObserver.next(navigator.onLine);
+        return () => {
+          window.removeEventListener(ConnectivityService.OFFLINE_EVENT, offlineEventListener);
+          window.removeEventListener(ConnectivityService.ONLINE_EVENT, onlineEventListener);
+        };
+      }
+    );
+
+    // Publish a multicast ReplaySubject that emits the most recent connectivity status
+    this.connectivityObserver = this.connectivityMonitor.pipe(publishReplay(1)) as ConnectableObservable<boolean>;
+    this.connectivityObserver.connect();
   }
 
   connectivity(): Observable<boolean> {
-    return this.subject.asObservable();
-    // TODO: Proper cleanup?
+    if (!this.connectivityObserver) {
```

Comment:
Your `providedIn: 'root'` should ensure this service is a singleton, so is this check needed?

Reply:
The conditional is related to the lazy initialisation rather than the single, app-wide instance that'll be instanced due to `providedIn: 'root'`.

Similar to comment above... thinking re-usability if this were to be provided in specific modules etc.

Reply:
Understood, although I think the responsibility of the lazy initialisation can be left to Rx, no?  

When you define an Observable, nothing is executed until it is first subscribed to (lazy evaluation), in which case maybe you can just define `connectedStatus` directly, resulting in more concise code?

``` typescript
export class ConnectivityService {
  connectedStatus = new Observable(...);
  connectivityObserver = this.connectedStatus.pipe(
    shareReplay({bufferSize: 1, refCount: true})
  ) as ConnectableObservable<boolean>;
}
```

Also, find `connectivityObserver` an unusual name.  We only need one of these properties public right, the one with the status.  If so, I'd keep the implementation-specific terminology private.


``` typescript
export class ConnectivityService {
  private connectionStateObservable = new Observable(...);
  connectionStatus = this.connectedStatus.pipe(
    shareReplay(1) // shorthand for previous object, I think
  );
}
```

Reply:
Googled it and also found this implementation too, lol

```
merge(
  fromEvent(window, 'offline').pipe(mapTo(false)),
  fromEvent(window, 'online').pipe(mapTo(true)),
  of(navigator.onLine),
);
```